### DISCUSSION TITLE
Before Migration a question according pusher-js / reconnection issues

### DISCUSSION THREAD

**Discussion by ibrainventures** [soketi/pws#314] [Q&A]
In my actual Webapp i am Using laravel-echo with socket.io 2.3 and on the backend laravel 8, and tladverdure laravel websocket on a dedicated server (redis, horizon for communication / eventing..)

Two Years go i tested the beyondcode laravel-php-based websocket server with the pusher-js connector instead of socket.io and there was a major issue with reconnection after the third disconnection from pusher-js to the websocket service.

After many, many hours of testing and workarounds the final conclusion was, that pusher-js is the culprit;

My Issue / Discussion to this issue:

https://github.com/beyondcode/laravel-websockets/issues/163

all those "i find a solution" where not stable and pusher-js was often calling "home" (pusher origin service) instead of 
the service websocket service.

So i switched back and the combination socket.io 2.3 <-> tladverdure  is working still stable.

Your soketi Server sounds really great, cudos!,  an bevor a big migration may you can test your stack
easy against this issue:

- Start a app which connects to your ws server inside your smartphone. 
- "switch off" (normal standby) the smartphone and wait some seconds / minutes (so the ws connection is closed)
- "switch on" the smartphone and look is the pusher-js / websocket reconnecting fine ....
- >> repeat 3 times

after doing this 3-times, pusher-js didnt reconnect to the websocket server automatic .. it "says" that
it is reconnecting, but with the wrong server ...

Has (your) solution - (or better say pusher-js still) the same issue ?  
If yes, is there a possibility to use socket.io (4) instead of the pusher-js ?

This would be great and many thanks.

**Comment by rennokki** [comment]
FWIK, Pusher's JS library is trying to reconnect exponentially to the server in case of unknown disconnections. These disconnections are not accounted for if they are server-based (for example, when soketi is closing, it sends the [`4200` code](https://pusher.com/docs/channels/library_auth_reference/pusher-websockets-protocol/#4200-4299) to reconnect immediately).

On transition to offline, you might want to listen for these events and try to disconnect and connect on Pusher automatically. In the meanwhile, the connection will still be opened for 120 more seconds in the server, but it will be ejected after this period of time.

I'm pretty sure it how you're handling the connection on the device using the Pusher SDK.

**Comment by ibrainventures** [comment]
Hi Renokki, thank you very much for the answer.

I self gave me a little fresh-up , what happens to this reconnect-issue since 2019 and saw
some investigations and solutions (also a forked-patched pusher 7. - only ws connection 
package) and also saw that culprit in the analyse of the "counting of 2 failed ws connections and it is using the
xhr connection" inside the pusher-js sourcecode, which is still used.

My feeling: to detect the transition or the interpreting / counting by pusher-js is not always given, as there
are also solutions outside with transition detect and a complete app-reload-process.

So my little question, (and hope), is there any chance to use on the frontend-side
the good-old **socket.io** v4 package (laravel-echo seems to working now ..) together with the soketi
Server? (in knowledge of losing some feature ..) or is this completly hopeless?

**Reply by rennokki** [reply]
Prior to soketi, [soketi/echo-server](https://github.com/soketi/echo-server) existed. It's Pusher but it works on Socket.IO. I won't recommend it though, as it might be buggy and it was deprecated/archived in favor of the nowadays soketi.

Socket.IO is not compatible with this app. :)
